File: ruby/abc219_d.rb
# https://atcoder.jp/contests/abc219/tasks/abc219_d

n = gets.chomp.to_i
x, y = gets.chomp.split(" ").map(&:to_i)
list_ab = n.times.map { gets.chomp.split(" ").map(&:to_i) }

total_x, total_y = 0, 0
list_ab.each { |a, b|
  total_x += a
  total_y += b
}
if total_x < x || total_y < y
  -1
  exit
end

i = 1
min_size = 500
while i < 2 ** n - 1
  bit = i.to_s(2).chars.map(&:to_i)
  size = bit.size
  cur_size = size - bit.count(&:zero?)
  sum_x, sum_y = 0, 0
  bit.each { |idx|
    next if idx == 0
    dx, dy = list_ab[size - idx]
    sum_x += dx
    sum_y += dy
  }
  if x < sum_x && y < sum_y && min_size > cur_size
    min_size = cur_size
  end
  i += 1
end

ans = if min_size == 500
  -1
else
  min_size
end
puts ans

File: rust/src/abc219_d.rs
// https://atcoder.jp/contests/abc219/tasks/abc219_d

use proconio::input;

fn main() {
    input! {
        n: usize,
        x: usize,
        y: usize,
        list: [(usize, usize); n]
    }

    let max = 500;

    let mut dp = vec![vec![vec![max; y + 1]; x + 1]; n + 1];
    dp[0][0][0] = 0;
    for i in 0..n {
        let (a, b) = list[i];
        for j in 0..(x+1) {
            for k in 0..(y+1) {
                let ux = x.min(a + j);
                let uy = y.min(b + k);

                // 購入
                dp[i + 1][ux][uy] = dp[i + 1][ux][uy].min(dp[i][j][k] + 1);
                // 購入しない
                dp[i + 1][j][k] = dp[i + 1][j][k].min(dp[i][j][k]);
            }
        }
    }

    let ans = if dp[n][x][y] == max {
        -1
    } else {
        dp[n][x][y]
    };
    println!("{}", ans);
}